opm/grid/GraphOfGridWrappers.hpp

```diff
@@ -0,0 +1,270 @@
+            *err = ZOLTAN_FATAL;
+            return;
+        }
+        for (const auto& e : eList)
+        {
+            nborGIDs[id]= e.first;
+            nborProc[id]= gog.getVertex(e.first).nproc;
+            edgeWeights[id]= e.second;
+            ++id;
+        }
+    }
+    *err = ZOLTAN_OK;
+}
+
+/// \brief Register callback functions to Zoltan
+template<typename Zoltan_Struct>
+void setGraphOfGridZoltanGraphFunctions(Zoltan_Struct *zz,
+                      const GraphOfGrid<Dune::CpGrid>& gog)
+{
+    GraphOfGrid<Dune::CpGrid>* pGraph = const_cast<GraphOfGrid<Dune::CpGrid>*>(&gog);
+    Zoltan_Set_Num_Obj_Fn(zz, getGraphOfGridNumVertices, pGraph);
+    Zoltan_Set_Obj_List_Fn(zz, getGraphOfGridVerticesList, pGraph);
+    Zoltan_Set_Num_Edges_Multi_Fn(zz, getGraphOfGridNumEdges, pGraph);
+    Zoltan_Set_Edge_List_Multi_Fn(zz, getGraphOfGridEdgeList, pGraph);
+}
+
+/// \brief Adds well to the GraphOfGrid
+///
+/// Translates wells' cartesian ID to global ID used in the graph.
+/// Adding the well contracts vertices of the well into one vertex.
+///
+/// checkWellIntersections==true makes the algorithm check if wells
+/// intersect and if their cell IDs are present in the graph.
+/// Setting it to false makes the algorithm faster but leaves user
+/// responsible for keeping wells disjoint.
+void addFutureConnectionWells (GraphOfGrid<Dune::CpGrid>& gog,
+    const std::unordered_map<std::string, std::set<int>>& wells,
+    bool checkWellIntersections=true)
+{
+    // create compressed lookup from cartesian.
+    const auto& grid = gog.getGrid();
+    const auto& cpgdim = grid.logicalCartesianSize();
+    std::vector<int> cartesian_to_compressed(cpgdim[0]*cpgdim[1]*cpgdim[2], -1);
+    for( int i=0; i < grid.numCells(); ++i )
+    {
+        cartesian_to_compressed[grid.globalCell()[i]] = i;
+    }
+
+    for (const auto& w: wells)
+    {
+        std::set<int> wellsgID;
+        for (const int& cell : w.second)
+        {
+            int gID = cartesian_to_compressed.at(cell);
+            assert(gID!=-1); // well should be an active cell
+            wellsgID.insert(gID);
+        }
+        gog.addWell(wellsgID,checkWellIntersections);
+    }
+}
+
+/// \brief Add well cells' global IDs to the list
+///
+/// Output of the partitioning is missing vertices that were contracted.
+/// This function fills in omitted gIDs and gives them the properties
+/// (like process number and ownership) of their representative cell (well ID).
+template<typename TheTuple>
+void extendImportExportList (const GraphOfGrid<Dune::CpGrid>& gog,
+                                       std::vector<TheTuple>& cellList)
+{
+    // using TheTuple = std::tuple<int,int,char>; or std::tuple<int,int,char,int>
+    using CellList = std::vector<TheTuple>;
+    // make a list of wells for easy identification. Contains ID, begin, end
+    using iter = std::set<int>::const_iterator;
+    std::list<std::tuple<int,iter,iter>> wellList;
```

Comment:
Please add a comment here why having this list is beneficial: Because we can make the search space smaller below.

We do a (linear) search in this container for the first cell below. Maybe we should use a vector sorted by the first cell? Then we could do  a binary search with std::lower_bound below.

We cannot shrink the container though, but maybe it is still more performant?

Could also be premature optimization (and turn out slower or not that much faster).IF we do such a change we should actually benchmark both on some cases to see which is better. It is not like we have thousands of wells...

So maybe this is for the future.

Reply:
I changed the container from std::list to std::unordered_map. Removing entries no longer improves search speed, but allows us to leave the loop sooner when the map gets empty.